src/sknnr/transformers/_cca.py

```diff
@@ -13,228 +22,85 @@ def zero_sum_columns(arr: NDArray) -> NDArray:
+        ...
+
+    def _check_inputs(self):
+        if not is_2d_numeric_numpy_array(self.X):
+            msg = "X must be a 2D numeric numpy array"
+            raise ValueError(msg)
 
-    @property
-    def V(self):
-        return self._V / np.sqrt(self.cw)[:, None]
+        if not is_2d_numeric_numpy_array(self.Y):
+            msg = "Y must be a 2D numeric numpy array"
+            raise ValueError(msg)
 
-    @property
-    def _WA(self):
-        x = np.diag(1.0 / np.sqrt(self.S))
-        return self.transform.Y.dot(self._V).dot(x)
+        if self.X.shape[0] != self.Y.shape[0]:
+            msg = "X and Y must have the same number of rows"
+            raise ValueError(msg)
 
-    @property
-    def WA(self):
-        return self._WA / np.sqrt(self.rw)[:, None]
+        if np.any(zero_sum_rows(self.X)):
+            msg = "All row sums in X must be greater than 0"
+            raise ValueError(msg)
 
-    @property
-    def biplot_scores(self):
-        a = 1.0 / np.sqrt(np.square(self.X_scale).sum(axis=0))
-        b = self.X_scale.T.dot(self._U)
-        return a[:, None] * b
+        excluded_columns = zero_sum_columns(self.Y)
+        self.Y = self.Y[:, ~excluded_columns]
 
-    @property
-    def Y_resid(self):
-        return self.transform.Y - self.Y_fit
+    def __call__(self):
+        X_scale = self._transform_X()
+        Q, R = np.linalg.qr(X_scale)
+        solution, _, rank, _ = np.linalg.lstsq(R, Q.T @ self.Y, rcond=None)
+        Y_fit = X_scale @ solution
+        U, S, _ = np.linalg.svd(Y_fit, full_matrices=False)
+        self.rank = min(rank, np.sum([S > self.ZERO]))
+        self.eigenvalues = np.square(S)[: self.rank]
+        self.axis_weights = np.diag(np.sqrt(self.eigenvalues / self.eigenvalues.sum()))
+        self.coefficients = np.linalg.lstsq(R, Q.T @ U[:, : self.rank], rcond=None)[0]
+        return self
 
     @property
     def max_components(self):
         return self.rank
 
-    @property
-    def eigenvalues(self):
-        return self.S
-
-    @property
-    def axis_weights(self):
-        return np.diag(np.sqrt(self.S / self.S.sum()))
-
-    @property
-    def coefficients(self):
-        Q, R = self.qr
-        right = Q.T @ self._U
-        return np.linalg.lstsq(R, right, rcond=None)[0]
-
     def projector(self, n_components):
         return (
             self.coefficients[:, :n_components]
             @ self.axis_weights[:n_components, :n_components]
         )
 
-    @property
-    def species_scores(self):
-        return np.multiply(self.V, np.sqrt(self.S))
-
-    @property
-    def site_lc_scores(self):
-        return self.U
 
-    @property
-    def site_wa_scores(self):
-        return self.WA
+class CCA(ConstrainedOrdination):
+    def _transform_Y(self):
+        normalized = self.Y / self.Y.sum()
+        self.rw = normalized.sum(axis=1)
+        self.cw = normalized.sum(axis=0)
+        rc = np.outer(self.rw, self.cw)
+        self.Y = (normalized - rc) / np.sqrt(rc)
```

Comment:
I'm a little surprised that `_transform_X` returns the transformed X while `_transform_Y` sets `Y` attribute. Is that because `_transform_Y` mutates the class and should only be called once during initialization, while `_transform_X` can get called repeatedly? I hate to suggest renaming without offering any options, but maybe renaming one of them would be an easy way to make it clear that they work differently?

Reply:
You're absolutely right on this.  For now, both `_transform_X` and `_transform_Y` resets the `self.X` and `self.Y` arrays within the functions.  I don't love this.  In addition to the side effects of setting these within these methods, we're also setting other instance attributes (e.g. `env_center`, `rw`, and `cw` in `CCA` and `env_center` in `RDA`).  I'd love your opinion on where it would be best to set these instance attributes.  Do you feel like it's OK to hide within these methods?

Reply:
Cool, I think this is an improvement. The side effects of those methods, and the fact they have to be called in the right order does make me a little uncomfortable, too, though.

Maybe one minor improvement could be to have the `_transform` methods take `X` and `Y` parameters, rather than using `self.X` and `self.Y` (although they would still *set* those attributes). That would prevent the possibility of somehow calling them twice and double-transforming the `X` and `Y`.

The other thought is that maybe they could be renamed to at least make it clear that they will have side-effects, e.g. `_set_transformed_X(X: np.NDArray) -> None`. That doesn't help the fact that they also set other attributes or that they need to be called in the right order, so I guess another option would be to combine the two methods into one method with a generic name like `_initialize_ordination` or `_set_initialization_attrs` or something, which would take `X` and `Y` params.

Reply:
Cool, I think this is an improvement. The side effects of those methods, and the fact they have to be called in the right order does make me a little uncomfortable, too, though.

Maybe one minor improvement could be to have the `_transform` methods take `X` and `Y` parameters, rather than using `self.X` and `self.Y` (although they would still *set* those attributes). That would prevent the possibility of somehow calling them twice and double-transforming the `X` and `Y`.

The other thought is that maybe they could be renamed to at least make it clear that they will have side-effects, e.g. `_set_transformed_X(X: np.NDArray) -> None`. That doesn't help the fact that they also set other attributes or that they need to be called in the right order, so I guess another option would be to combine the two methods into one method with a generic name like `_initialize_ordination` or `_set_initialization_attrs` or something, which would take `X` and `Y` params.

Reply:
Great points, here.  I'll noodle on this, but definitely passing both `X` and `Y` together make much more sense.  So, I assume you're thinking about something like:
```
class ConstrainedOrdination(ABC):
    def __init__(self, X, Y):
        X = np.asarray(X)
        Y = np.asarray(Y)
        self._check_inputs(X, Y)
        self._set_initialization_attrs(X, Y) # this would be @abstractmethod
```

Reply:
Yeah, that was what I had in mind.

Reply:
I've made this change.  The one change I'm not thrilled with is that I'm passing `X` and `Y` back from `self._check_inputs` because `Y` can be modified based on if there are zero-sum columns.  Even though there is no chance that `X` gets modified, it felt weird just returning `Y` and I didn't want to set `self.Y` in `_check_inputs`.  Let me know what you think of this.  The other thing I just realized is that we can make `_check_inputs` a `staticmethod`.

Reply:
Yeah, I can see how it's a little odd to return `X` unchanged, but I agree with your logic that the alternatives would be weirder. This looks good to me!